app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt

```diff
@@ -221,17 +226,96 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
   }
 
   fun onOngoingStoryClicked(promotedStory: PromotedStory) {
-    playExploration(promotedStory.topicId, promotedStory.storyId, promotedStory.explorationId)
+    val shouldSavePartialProgress =
+      when (promotedStory.chapterPlayState) {
+        ChapterPlayState.IN_PROGRESS_SAVED, ChapterPlayState.IN_PROGRESS_NOT_SAVED,
+        ChapterPlayState.STARTED_NOT_COMPLETED, ChapterPlayState.NOT_STARTED -> true
+        else -> false
+      }
+    if (promotedStory.chapterPlayState == ChapterPlayState.IN_PROGRESS_SAVED) {
+      explorationCheckpointController.isSavedCheckpointCompatibleWithExploration(
+        ProfileId.getDefaultInstance(),
+        promotedStory.explorationId
+      ).toLiveData().observe(
+        fragment,
+        Observer {
+          if (it.isSuccess()) {
+            startOrResumeExploration(
+              internalProfileId,
+              promotedStory.topicId,
+              promotedStory.storyId,
+              promotedStory.explorationId,
+              shouldSavePartialProgress,
+              canExplorationBeResumed = it.getOrThrow(),
+              backflowScreen = null
+            )
+          } else if (it.isFailure()) {
+            startOrResumeExploration(
+              internalProfileId,
+              promotedStory.topicId,
+              promotedStory.storyId,
+              promotedStory.explorationId,
+              shouldSavePartialProgress,
+              canExplorationBeResumed = false,
+              backflowScreen = null
+            )
+          }
+        }
+      )
+    } else {
+      startOrResumeExploration(
+        internalProfileId,
+        promotedStory.topicId,
+        promotedStory.storyId,
+        promotedStory.explorationId,
+        shouldSavePartialProgress,
+        canExplorationBeResumed = false,
+        backflowScreen = null
+      )
+    }
   }
```

Comment:
There are two erros due to this code:

1. If the exploration version doesn't match, both resume lesson and exploration activity are opened one after the other, when only one should have been opened.


I think the reason because exploration Activity and ResumeLessonActivity is starting are starting (one after the other) if the exploration doesn't match is since when the exploration versions don't match we start first call play exploration (that loads exploration and then saved checkpoint) and then open exploration activity, in the mean time version of exploration is updated (because play exploration saves a new checkpoint), the observer is notified and this time resume exploration activity is opened.




2. This error will surely not occur in real-world scenarios so I am not sure that if we should take care of this: if the exploration version is updated while the user is in the exploration (cannot occur in real-world scenarios), when the user clicks back to go back to the topic lessons fragment, upon returning to the topic lessons fragment they are automatically taken back to the exploration.

I am not sure why this is happening, but I think it might be because the version is now updated so the observer (that matched the version in the topic lessons fragment) starts the exploration again.

Reply:
My suggestion elsewhere to change the flow to navigate based on the state of the checkpoint DataProvider might solve this (presuming we're doing that in ExplorationFragment) since we wouldn't be passing in an actual checkpoint to initiate the exploration if it's incompatible. Merging the two DataProviders together should avoid the race condition described here.

Reply:
as per chats in hangouts, used removeObserver.